File: server/src/utils/policy.ts
import type { Policy } from "../types.js";
import { nowUnix } from "./ids.js";

export const DEFAULT_POLICY = (updatedBy = "system"): Policy => ({
  allowText: true,
  allowImage: true,
  allowFile: true,
  maxFileSizeBytes: 3 * 1024 * 1024,
  version: 1,
  updatedBy,
  updatedAtUnix: nowUnix()
});

export const isItemAllowedByPolicy = (
  policy: Policy,
  kind: "text" | "image" | "file",
  sizeBytes: number
): boolean => {
  if (kind === "text") {
    return policy.allowText;
  }
  if (kind === "image") {
    return policy.allowImage;
  }
  if (!policy.allowFile) {
    return false;
  }
  return sizeBytes < policy.maxFileSizeBytes;
};

File: crates/client-runtime/src/policy.rs
use crate::types::{ClipboardKind, ClipboardPayload, SharePolicy};

pub const DEFAULT_MAX_FILE_SIZE_BYTES: usize = 3 * 1024 * 1024;

pub fn default_policy() -> SharePolicy {
    SharePolicy {
        allow_text: true,
        allow_image: true,
        allow_file: true,
        max_file_size_bytes: DEFAULT_MAX_FILE_SIZE_BYTES,
        version: 1,
    }
}

pub fn is_allowed_by_policy(policy: &SharePolicy, payload: &ClipboardPayload) -> bool {
    match payload.kind {
        ClipboardKind::Text => policy.allow_text,
        ClipboardKind::Image => policy.allow_image,
        ClipboardKind::File => policy.allow_file && payload.size_bytes < policy.max_file_size_bytes,
    }
}
